### ISSUE TITLE
Issues with Retina-screen, createGraphics and pixeldensity

### ISSUE THREAD

**Issue by rockfarmor** [processing/p5.js#2077]
Hello! I've been looking around the forum for similar issues, and found some that partly handles my problem. It's known that when using createGraphics() on a screen with pixeldensity > 1, the graphicsObject becomes distorted. A known fix for this is to manually set the pixeldensity of the graphicsObject to 1. But now to the problem -> When trying to render text to the object, the text-quality is becomes really low, and when rendering the text with the graphicsObject's pixeldensity=2 the text becomes distorted and to large.

#### Nature of issue?

- [x] Found a bug
- [x] Existing feature enhancement
- [ ] New feature request

#### Most appropriate sub-area of p5.js?

- [x] Color
- [x] Core
- [ ] Events
- [x] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL

#### Which platform were you using when you encountered this?

- [ ] Mobile/Tablet (touch devices)
- [x] Desktop/Laptop
- [x] Macbook with retina-screen
- [ ] Others (specify if possible)

#### Details about the bug: 
- p5.js version: 0.5.11
- Web browser and it's version: Chrome
- Operating System: Mac OSX
- Steps to reproduce this: Rendering text to a graphics-object with pixelDensity 1, on a screen with pixelDensity 2

Here are two printscreens displaying the problem. The text "Numbers:" is rendered to the GraphicsObject with very low resolution, whilst the text "windowtitle" is high-res and rendered to the canvas directly.

GraphcisObject with density 1 (On screen with density2)
<img width="368" alt="2" src="https://user-images.githubusercontent.com/10855270/28295164-2af21268-6b5f-11e7-9de1-3db7a6c4faae.png">

GraphcisObject with density 2 (On screen with density2)
<img width="358" alt="1" src="https://user-images.githubusercontent.com/10855270/28295160-25eea72c-6b5f-11e7-81ab-eaf5f234065e.png">

Tldr; Is there any work-around to render high-res text to the p5.Graphics object on high-pixelDensity screens

**Comment by limzykenneth** [comment]
May not be helpful but is it possible to use DOM element instead? By using DOM element to render text, the browser and in turn the OS will take care of text on different pixel density display so you don't need to worry about this and let the canvas do graphics drawing instead of text rendering.

**Comment by quinton-ashley** [comment]
yeah but what if you want to manipulate text at a high resolution in p5?  I'm sure I'm not just speaking for myself when I say this issue shouldn't be ignored.  Of course you can render OP's program in html but that's not the point of Processing.

**Comment by limzykenneth** [comment]
Not saying anything should be ignored. Just suggesting possible alternatives. If there's an easier workaround perhaps it could help. 

Of course depending on use case that may not be possible but I'm not sure if it is the case for the OP's program so I'm just throwing the suggestion out there.

**Comment by quinton-ashley** [comment]
Oh okay sorry I see where you're coming from now.

**Comment by rockfarmor** [comment]
Thanks for the feedback! Unfortunately I can't use a DOM element for this program. The purpose of using the graphics object is to create a "text-box" of width and height where the user can scroll through the content of the box, where the text "outside" the box won't be visible. This currently works with the graphics object, but the resolution on a screen with density 2 is freaking terrible.

**Comment by lmccart** [comment]
as a temporary fix, maybe it would work to up the pixelDensity of the graphics object, then scale it down so it's the correct size?

```javascript
var g = createGraphics(400, 400);
g.pixelDensity(4);
g.scale(1/4);
```

**Comment by rockfarmor** [comment]
I've tried this before, it scales the graphics object right, but still produces the same "low-res" result.

**Comment by Spongman** [comment]
here's another repro for this problem i just ran into: https://codepen.io/Spongman/pen/zpqVpq?editors=0011

the white circles should be the same size, but they're not because the size calculation for `createGraphics` is messed up wrt. `pixelDensity`. the upshot is that _no_ sketch that uses `createGraphics` will work correctly if `pixelDensity <> 1`.

IMO I believe the high-dpi scaling stuff should be disabled until it works correctly in all cases. a method should be available to enable it for sketches that don't care about the effects of this bug.

**Comment by n1ckfg** [comment]
I've been using this to fix coordinates inside graphics objects; haven't tested extensively on non-Retina screens but so far it seems to work well:
```
function fixGraphicsX(_x, _pg, _offset) {
	return map(_x, 0, _pg.width, 0, _pg.width/pixelDensity()) - (_offset/pixelDensity());
}

function fixGraphicsY(_y, _pg, _offset) {
	return map(_y, 0, _pg.height, 0, _pg.height/pixelDensity()) - (_offset/pixelDensity());
}
```

**Comment by BenjaminHabert** [comment]
Note that you can use `copy()` instead of `image()` to draw a Graphics object on the main canvas: it will scale properly. See [this example](https://alpha.editor.p5js.org/BenjaminHabert/sketches/ryKCVdK9G).

By the way: I do not understand why both `image()` and `copy()` are implemented. Is there anything that only one of those can do?

<img width="340" alt="graphics_copy_and_image" src="https://user-images.githubusercontent.com/9921832/38053337-1d12574c-32d4-11e8-986d-ce1c1cec94e1.png">

**Comment by BenjaminHabert** [comment]
I obtain the desired result with the fix proposed in the merge pull request.

<img width="310" alt="graphics_copy_and_image_fixed" src="https://user-images.githubusercontent.com/9921832/38055222-66b1ed2c-32d9-11e8-846b-bb9a9f964849.png">
